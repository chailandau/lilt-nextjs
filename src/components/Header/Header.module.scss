.header {
    flex-direction: row !important;
    padding-top: rem(24);
    padding-bottom: rem(16);
    margin-bottom: rem(-16);
    position: fixed;
    top: 0;
    z-index: 99999;
    width: 100vw;
    @include transition(padding-top, padding-bottom);
    @include paddingX(0);
    &:before {
        display: block;
        content: '';
        position: absolute;
        width: 100%;
        height: 100%;
        top: 0;
        left: 0;
        z-index: 101;
        background: color(dark, blue);
    }
    &__content {
        flex-direction: row;
        gap: rem(40);
        justify-content: space-between;
        align-items: center;
    }
    .logo,
    .right-content {
        z-index: 101;
    }
    .logo {
        width: rem(250);
        padding: unset;
        @include transition(width);
        &:before {
            display: none;
        }
    }
    .right-content {
        align-items: center;
        gap: rem(24);
        @include transition(gap);
        @include laptop-lg {
            gap: rem(32);
        }
        @include desktop-sm {
            gap: rem(44);
        }
        .tablet-btn {
            display: none;
            text-align: center;
            @include transition(padding, background-position);

            @include tablet {
                display: block;
            }
        }
    }
    &__sticky {
        @include paddingY(4);
        @include tablet {
            @include paddingY(12);
        }
        &:before {
            box-shadow: shadow(lightFuchsia);
            border-bottom: 1px solid color(neutral, grayBlue);
        }
        .logo {
            width: rem(180);
        }
        .tablet-btn {
            padding: rem(10) rem(15);
        }
        &.header__menu-open {
            padding-top: rem(24);
            padding-bottom: rem(16);
            &:before {
                display: none;
            }
            .logo {
                width: rem(250);
            }
        }
    }
}
