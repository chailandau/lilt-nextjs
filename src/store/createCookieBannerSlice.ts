import { ZuSlice } from 'declarations';

export interface CookieBannerSlice {
    showCookieBanner: boolean;
    setShowCookieBanner: (arg: CookieBannerSlice['showCookieBanner']) => void;
}

const createCookieBannerSlice: ZuSlice<CookieBannerSlice> = (
    set: (arg: () => Partial<CookieBannerSlice>) => void
) => ({
    showCookieBanner: false,
    setShowCookieBanner: (arg) => set(() => ({ showCookieBanner: arg }))
});

export default createCookieBannerSlice;
