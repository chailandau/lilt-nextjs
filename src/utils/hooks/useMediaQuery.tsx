import { useEffect, useMemo, useState } from 'react';

/**
 * Check if viewport matches provided media query
 *
 * https://developer.mozilla.org/en-US/docs/Web/API/Window/matchMedia
 * https://developer.mozilla.org/en-US/docs/Web/API/MediaQueryList/change_event
 *
 * Example:
 *
 *    useMediaQuery('(max-width: 600px)');
 *
 *    useMediaQuery('screen and (min-width: 600px)');
 */

export const useMediaQuery = (mediaQueryString: string): boolean => {
    if (typeof window !== 'undefined') {
        const query = useMemo(
            () => window.matchMedia(mediaQueryString),
            [mediaQueryString]
        );
        const [matches, setMatches] = useState(query.matches);
        useEffect(() => {
            const listener = (e: {
                matches: boolean | ((prevState: boolean) => boolean);
            }) => setMatches(e.matches);
            query.addEventListener('change', listener);

            return () => query.removeEventListener('change', listener);
        }, [query]);

        return matches;
    }

    return false;
};
