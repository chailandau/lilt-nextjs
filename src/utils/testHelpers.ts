import { fireEvent, render } from '@testing-library/react';
import { axe } from 'jest-axe';
import { ReactElement, cloneElement } from 'react';

type Component = ReactElement;

interface TestComponent {
    component: Component;
}

interface TestRole extends TestComponent {
    role: string;
}

interface TestTag extends TestComponent {
    htmlTag: string;
}

interface TestPropOptions extends TestTag {
    propName: string;
    propOptions: readonly string[];
}

interface TestTags extends TestComponent {
    tags: readonly string[];
}

interface TestRenderText extends TestRole {
    text: string;
}

/**
 * Maps through a list of prop options from a provided array, testing that each rendered component contains the specified prop option.
 *
 * @param component - Component to test.
 * @param propName - Name of prop to pass options to.
 * @param propOptions - Array of prop options to map through.
 * @param htmlTag - HTML tag of the component.
 */
export const testPropOptions = ({
    component,
    propName,
    propOptions,
    htmlTag
}: TestPropOptions): void => {
    propOptions.forEach((propOption) => {
        it(`has '${propOption}' class`, () => {
            const { container } = render(
                cloneElement(component, { [propName]: propOption })
            );
            expect(container.querySelector(htmlTag)).toHaveClass(propOption);
        });
    });
};

/**
 * Tests that `className` is applied to component if it exists.
 *
 * @param component - Component to test.
 * @param htmlTag - HTML tag of the component.
 */
export const testAppliesClass = ({ component, htmlTag }: TestTag): void => {
    it('applies class if it exists', () => {
        const { container } = render(
            cloneElement(component, { className: 'test' })
        );
        expect(container.querySelector(htmlTag)).toHaveClass('test');
    });
};

/**
 * Maps through a list of tags from a provided array, testing that each rendered component
 * contains the specified tag.
 *
 * @param component - Component to render.
 * @param tags - Array of tags to map through.
 */
export const testTags = ({ component, tags }: TestTags): void => {
    tags.forEach((tag) => {
        it(`has '${tag}' tag`, () => {
            const { container } = render(cloneElement(component, { as: tag }));
            expect(container.querySelector(tag)).toBeDefined();
        });
    });
};

/**
 * Tests that component renders with correct text.
 *
 * @param component - Component to render.
 * @param text - Text to render.
 */
export const testRenderText = ({ component, text }: TestRenderText): void => {
    it('renders with correct text', () => {
        const { container } = render(component);
        expect(container.textContent).toBe(text);
    });
};

/**
 * Tests that children render correctly inside a Higher Order Component (HOC) component.
 *
 * @param component - Component to test.
 */
export const testRenderChildren = ({ component }: TestComponent): void => {
    it('renders children inside HOC component', () => {
        const { getByText } = render(component);
        expect(getByText('Test')).toBeDefined();
    });
};

/**
 * Tests that keydown event is fired when key is pressed.
 *
 * @param component - Component to test.
 * @param role - Component's ARIA role.
 */
export const testKeyDown = ({ component, role }: TestRole): void => {
    it('should call onKeyDown when a key is pressed', () => {
        const onKeyDownMock = jest.fn();

        const { getByRole } = render(
            cloneElement(component, {
                onKeyDown: onKeyDownMock
            })
        );
        const element = getByRole(role);
        fireEvent.keyDown(element, { key: 'Enter', code: 13 });
        expect(onKeyDownMock).toHaveBeenCalledTimes(1);

        const eventObject = onKeyDownMock.mock.calls[0][0];
        expect(eventObject).toBeDefined();
        expect(eventObject.key).toBe('Enter');
    });
};

/**
 * Tests that component matches snapshot and does not have axe violations.
 *
 * @param component - Component to test.
 */
export const testAxeAndSnapshot = ({ component }: TestComponent) => {
    it('matches snapshot', () => {
        const { asFragment } = render(component);
        expect(asFragment()).toMatchSnapshot();
    });
    it('does not have axe violations', async () => {
        const { container } = render(component);
        expect(await axe(container)).toHaveNoViolations();
    });
};
