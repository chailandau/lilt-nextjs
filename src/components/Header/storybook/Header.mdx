import { Canvas, Meta, ArgTypes } from '@storybook/blocks';

import * as HeaderStories from './Header.stories';

<Meta of={HeaderStories} />

# Header Component

<Canvas of={HeaderStories.Default} />

The `Header` component accepts a `menuItems` array and a `callToAction` object.

## PayloadCMS

From dashboard, navigate to `Global` > `Header`.

### menuItems

Menu items contain the following fields (these vary depending on which `Menu Item Type` is selected):

-   `Menu Item Type` - select whether menu item is an internal/external
    link or a submenu
-   `Internal Link` - select site page to link to
    -   `Override Page Name` - override site page name with custom label
    -   `Custom Label` - input internal menu item's custom label
-   `External Link` - input a valid external URL - `Label` - menu item's label
-   `Submenu` - create nested menu item.
    -   Contains same options as regular menu item, except for submenu option (one level of submenu is currently supported).

<img
    src='docs/header/menu-items.gif'
    alt='walkthrough of how to set header menu items in PayloadCMS'
/>

<hr />

### callToAction

    - `Call To Action` - select the header's call to action button via the dropdown
        - Content editors can select an existing CTA or create a new one by clicking the `+` sign to the right of the dropdown.
        - See `Button` component [docs](?path=/docs/components-button--docs#buttondata) for a walkthrough on creating a Button in PayloadCMS.

<hr />

<img
    src='docs/header/cta.gif'
    alt='walkthrough of how to set header call to action in PayloadCMS'
/>

<br />

### Props Table

<ArgTypes />
