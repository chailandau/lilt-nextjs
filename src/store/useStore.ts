import { create } from 'zustand';

import createCookieBannerSlice, {
    CookieBannerSlice
} from './createCookieBannerSlice';

import createMenuSlice, { MenuSlice } from '@/store/createMenuSlice';

type StoreState = MenuSlice & CookieBannerSlice;

const useStore = create<StoreState>()((...args) => ({
    ...createMenuSlice(...args),
    ...createCookieBannerSlice(...args)
}));

export default useStore;
