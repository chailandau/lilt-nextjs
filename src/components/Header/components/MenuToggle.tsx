import classNames from 'classnames';

import styles from './MenuToggle.module.scss';

import Container from '@/atoms/Container/Container';
import useStore from '@/store/useStore';

const MenuToggle = () => {
    const { menuOpen, setMenuOpen } = useStore();

    const classList = classNames(
        styles['menu-toggle'],
        menuOpen && styles['open']
    );

    return (
        <Container
            as='button'
            className={classList}
            onClick={() => setMenuOpen(!menuOpen)}
            ariaLabel={'menu toggle'}
        >
            <Container className={styles['menu-toggle-container']}>
                <Container className={styles['menu-toggle-inner']}></Container>
            </Container>
        </Container>
    );
};

export default MenuToggle;
