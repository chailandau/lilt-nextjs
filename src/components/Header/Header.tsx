import FocusTrap from 'focus-trap-react';
import { FC, useEffect, useState } from 'react';

import Button from '../Button/Button';

import DesktopNav from './components/DesktopNav';
import MenuToggle from './components/MenuToggle';
import MobileNav from './components/MobileNav';
import styles from './Header.module.scss';

import { Button as ButtonType, Header_MenuItems } from '@/api/graphqlTypes';
import Logo from '@/assets/svg/Logo.svg';
import Image from '@/atoms/Image/Image';
import Link from '@/atoms/Link/Link';
import Flex from '@/molecules/Flex/Flex';
import Section from '@/molecules/Section/Section';
import useStore from '@/store/useStore';
import { useMediaQuery } from '@/utils/hooks/useMediaQuery';

export interface HeaderProps {
    /* Menu items to display */
    menuItems: Header_MenuItems[] | null;
    /* CTA object */
    callToAction?: ButtonType | null;
}
const Header: FC<HeaderProps> = ({ menuItems, callToAction }) => {
    const { menuOpen } = useStore();
    const [headerEl, setHeaderEl] = useState<HTMLElement | null>(null);

    const isLaptop = useMediaQuery('(min-width: 992px)');

    useEffect(() => {
        setHeaderEl(
            document.querySelector(`.${styles['header']}`) as HTMLElement
        );
    }, []);

    return (
        <>
            <FocusTrap
                active={menuOpen}
                containerElements={headerEl ? [headerEl] : []}
            />
            <Section as='header' className={styles['header']}>
                <Link
                    href={process.env.NEXT_PUBLIC_BASE_URL as string}
                    className={styles['logo']}
                >
                    <Image src={Logo} alt='Long Island Laser Tag' priority />
                </Link>

                <Flex className={styles['right-content']}>
                    {isLaptop && (
                        <DesktopNav
                            menuItems={menuItems}
                            callToAction={callToAction}
                        />
                    )}
                    {callToAction && (
                        <Button
                            buttonData={callToAction}
                            className={styles['tablet-btn']}
                        />
                    )}
                    <MenuToggle />
                </Flex>
                {!isLaptop && (
                    <MobileNav
                        menuItems={menuItems}
                        callToAction={callToAction}
                    />
                )}
            </Section>
        </>
    );
};

export default Header;
