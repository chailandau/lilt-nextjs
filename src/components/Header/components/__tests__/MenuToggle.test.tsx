/** @jest-environment jsdom */

import { fireEvent, render } from '@testing-library/react';

import Header from '../../Header';
import MenuToggle from '../MenuToggle';

import { menuItems } from '@/__mocks__/Header.mock';
import { mockUseMediaQuery } from '@/__mocks__/useMediaQuery.mock';
import * as useStoreModule from '@/store/useStore';
import { testAxeAndSnapshot } from '@/utils/testHelpers';

jest.mock('@/store/useStore', () => ({
    __esModule: true,
    ...jest.requireActual('@/store/useStore')
}));

mockUseMediaQuery(false);

describe('Menu Toggle', () => {
    it('opens navigation on click', async () => {
        const setMenuOpen = jest.fn();

        jest.spyOn(useStoreModule, 'default').mockReturnValue({
            menuOpen: false,
            setMenuOpen
        });

        const { findByRole } = render(<Header menuItems={menuItems} />);

        const button = await findByRole('button');

        fireEvent(button, new MouseEvent('click', { bubbles: true }));

        expect(setMenuOpen).toHaveBeenCalledWith(true);
        expect(button).toHaveClass('open');
    });

    testAxeAndSnapshot({ component: <MenuToggle /> });
});
