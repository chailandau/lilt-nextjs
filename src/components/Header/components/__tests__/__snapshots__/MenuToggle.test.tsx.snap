// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu Toggle matches snapshot 1`] = `
<DocumentFragment>
  <button
    aria-label="menu toggle"
    class="menu-toggle"
  >
    <div
      class="menu-toggle-container"
    >
      <div
        class="menu-toggle-inner"
      />
    </div>
  </button>
</DocumentFragment>
`;
